### PR TITLE
[ALLUXIO-2206] Fix a bug in LocalFirstPolicy and RoundRobinPolicy

### PR DESCRIPTION
**Pull request by peisun1115**

### CODE REVIEW

**Comment by gpang** [comment]
@peisun1115 thanks for the fix! Could you create a ticket for this?

**Comment by yupeng9** [comment]
I don't think this is a bug. For local first, when tiered storage is available and MEM availability is low, we still want to choose the local and evict stale blocks.

**Comment by peisun1115** [comment]
I think the available capacity includes tiered storage capacity.  See MasterWorkerInfo.java & line=131 .

**Comment by yupeng9** [comment]
Sure. I meant even without tiered storage, we can still evict and discard old blocks in favor for the new ones. @apc999

**Comment by peisun1115** [comment]
OK. If that is the case, then it is probably better change the default policy to be a different one? I feel it is confusing to the user when it fails to write to Alluxio though Alluxio has enough capacity. When LocalFirstPolicy is used, Alluxio basically becomes a local buffer cache because it is likely that every worker has enough total capacity to hold a block.

File: core/client/src/main/java/alluxio/client/block/BlockWorkerInfo.java [hunk 1]
```diff
@@ -56,6 +56,13 @@ public long getCapacityBytes() {
   }
 
   /**
+   * @return the remaining capacity of the worker in bytes
+   */
+  public long getAvailableCapacityBytes() {
```
Comment (apc999) [inline core/client/src/main/java/alluxio/client/block/BlockWorkerInfo.java]: NIT: maybe just `getAvailableBytes`?

**Comment by apc999** [comment]
IMO, a write policy might simply be the order of the following four:

(1) Writing to Local without Eviction
(2) Writing to Local with Eviction
(3) Writing to Remote without Eviction
(4) Writing to Remote with Eviction

And the current local-first realizes (1)&(2) > (3) > (4)
and by this PR, it makes local-first to be (1) > (3) > (4)

I guess there is no right or wrong answer to the policy---both may have legit use cases.
Should we probably just come up with a richer set of policies for users to configure?

**Comment by peisun1115** [comment]
I think the current implementation is (1) = (2) > (3) = (4) . 
By this PR, (1) > (3) > (2) > (4)

I can keep the current implementation and come up with a different policy.  
a. I think it is better to change the default policy. How do you think?
b. I feel the current one is less useful especially when the data is in remote object store such as s3 (the usecase we target most?) . In that sense, maybe we should remove that to make it less confusing. But i have no problem with keeping it there.

**Comment by yupeng9** [comment]
I think we should add more policies, and then we can discuss which one should be default. These can be separate tasks.

**Comment by apc999** [comment]
+1 for what @yupeng9  proposes. @peisun1115 does that sound good to you?
Also, we should have meaningful names and sufficient docs on these policies (including existing ones)

**Comment by peisun1115** [comment]
@yupeng9 @apc999 I can close this PR for now given that you two don't think this is a bug and work on adding more policies later.

**Pull request closed without merging**